src/index.js

```diff
@@ -55,59 +55,62 @@ const createRunner = ({ runtime = "worker_threads" } = {}) =>
      * @param {*} onResult
      * @param {*} onFailure
      */
-    runTests(tests, watcher, onStart, onResult, onFailure) {
+    async runTests(tests, watcher, onStart, onResult, onFailure) {
       const pool = this._pool;
       const { updateSnapshot, testNamePattern, maxWorkers } = this._config;
-
-      if (!this._runInBand && this._isProcessRunner) {
-        return pMap(
-          tests,
-          test => {
-            onStart(test);
-
-            return pool.run({ test, updateSnapshot, testNamePattern }).then(
-              result => void onResult(test, result),
-              error => void onFailure(test, error),
-            );
-          },
-          { concurrency: maxWorkers },
-        ).finally(async () => {
-          for (const { process } of pool.threads) {
-            // Use `process.disconnect()` instead of `process.kill()`, so we can collect coverage
-            // See https://github.com/nicolo-ribaudo/jest-light-runner/issues/90#issuecomment-2812473389
-            // Only override the first call https://github.com/tinylibs/tinypool/blob/dbf6d74282dd6031df8fc5c7706caef66b54070b/src/runtime/process-worker.ts#L61
-            const originalKill = process.kill;
-            process.kill = signal => {
-              if (!signal) {
-                process.disconnect();
-                process.kill = originalKill;
-                return;
-              }
-              return originalKill.call(process, signal);
-            };
-          }
-
-          await pool.destroy();
-        });
-      }
-
-      return Promise.all(
-        tests.map(test => {
+      const isProcessRunner = !this._runInBand && this._isProcessRunner;
+
+      let run;
+      if (isProcessRunner) {
+        run = async test => {
+          await onStart(test);
+          return pool.run({ test, updateSnapshot, testNamePattern });
+        };
+      } else {
+        run = test => {
           const mc = new MessageChannel();
           mc.port2.onmessage = () => onStart(test);
           mc.port2.unref();
 
-          return pool
-            .run(
-              { test, updateSnapshot, testNamePattern, port: mc.port1 },
-              { transferList: [mc.port1] },
-            )
-            .then(
-              result => void onResult(test, result),
-              error => void onFailure(test, error),
-            );
-        }),
+          return pool.run(
+            { test, updateSnapshot, testNamePattern, port: mc.port1 },
+            { transferList: [mc.port1] },
+          );
+        };
+      }
+
+      const mutex = pLimit(maxWorkers);
+
+      await Promise.all(
+        tests.map(test =>
+          mutex(async () => {
+            try {
+              onResult(test, await run(test));
+            } catch (error) {
+              onFailure(test, error);
+            }
```

Comment:
Since we used `await onStart(test)`, maybe use `await` on them too? I took a quick look at the jest code and it seems that they all return `Promise`.

Reply:
I didn't realize it's a promise, let's do this then 😄 

[`150a890` (#95)](https://github.com/nicolo-ribaudo/jest-light-runner/pull/95/commits/150a890f979bfd12843a779635b7d6a41b8e815f)